### PR TITLE
[POAE7-656] porting dump-all-pages-to-PMem solution in memory spill framework

### PR DESCRIPTION
**Pull request by yma11**
### What changes were proposed in this pull request?
porting dump-all-pages-to-PMem solution in memory spill framework

### Why are the changes needed?
unify the framework for different spill strategy.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
UT

### CODE REVIEW

**Comment by yma11** [comment]
@winningsix @XinDongIntel , Please take a look. This PR not include concurrent sort/write and PMem read buffer. Will submit separate PRs.

File: core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java [hunk 1]
```diff
@@ -421,6 +421,72 @@ public void freePage(MemoryBlock page, MemoryConsumer consumer) {
...
+  public MemoryBlock allocatePMemPage(long size) {
+    if (size > MAXIMUM_PAGE_SIZE_BYTES) {
+      throw new TooLargePageException(size);
+    }
+    final int pageNumber;
+    synchronized (this) {
+      pageNumber = allocatedPages.nextClearBit(0);
+      if (pageNumber >= PAGE_TABLE_SIZE) {
+        releaseExtendedMemory(size);
+        throw new IllegalStateException(
```
Comment (winningsix) [inline core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java:437]: Do we need to throw an exception or handle it (e.g. fallback or double size with warnning log)? Please address it in another PR.

File: core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java [hunk 2]
```diff
@@ -421,6 +421,72 @@ public void freePage(MemoryBlock page, MemoryConsumer consumer) {
...
+    try {
+      page = memoryManager.getUsablePMemPage(size);
+      if (page == null) {
+        page = memoryManager.extendedMemoryAllocator().allocate(size);
+        memoryManager.addPMemPages(page);
+      } else {
+        logger.debug("reuse pmem page.");
+      }
+    } catch (OutOfMemoryError e) {
+      logger.error("Failed to allocate a PMem page ({} bytes).", size);
```
Comment (winningsix) [inline core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java:452]: Throw this exception.

File: core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java [hunk 3]
```diff
@@ -421,6 +421,72 @@ public void freePage(MemoryBlock page, MemoryConsumer consumer) {
...
+      return null;
+    }
+    page.isExtendedMemory(true);
+    page.pageNumber = pageNumber;
+    pageTable[pageNumber] = page;
+    if (logger.isTraceEnabled()) {
+      logger.trace("Allocate page number {} ({} bytes)", pageNumber, size);
+    }
+    return page;
+
```
Comment (winningsix) [inline core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java:464]: remote this.

File: core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java [hunk 4]
```diff
@@ -421,6 +421,72 @@ public void freePage(MemoryBlock page, MemoryConsumer consumer) {
...
+      } else {
+        logger.debug("reuse pmem page.");
+      }
+    } catch (OutOfMemoryError e) {
+      logger.error("Failed to allocate a PMem page ({} bytes).", size);
+    }
+    if (page == null) {
+      return null;
+    }
+    page.isExtendedMemory(true);
```
Comment (winningsix) [inline core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java:457]: can we do this when allocating it?

File: core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PMemWriter.java [hunk 5]
```diff
@@ -104,12 +149,38 @@ private void updateLongArray(int numRecords, int position) {
...
+        }
+    }
+
+    public void clearAll() {
+        freeAllPMemPages();
     }
 
     @Override
     public int recordsSpilled() {
         //todo: implements recordsSpill here.
```
Comment (winningsix) [inline core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PMemWriter.java:175]: Remove TODO?

File: core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PMemWriter.java [hunk 6]
```diff
@@ -104,12 +149,38 @@ private void updateLongArray(int numRecords, int position) {
...
+        writeMetrics.incBytesWritten(pMemBlock.size());
+        this.sortedArray = new LongArray(pMemBlock);
+    }
+
+    @Override
+    public UnsafeSorterIterator getSpillReader() throws IOException {
+        if (diskSpillWriter != null) {
+            return diskSpillWriter.getSpillReader();
+        } else {
+            return new PMemReaderForUnsafeExternalSorter(sortedArray, position, totalRecordsWritten, taskMetrics);
```
Comment (winningsix) [inline core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PMemWriter.java:165]: how about the case we have partially? We can add a TODO here and address it in the future.

**Review by winningsix — COMMENTED** [review]
Just some minor comments.

File: core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PMemWriter.java (hunk 6 above)
Comment (yma11) [reply core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PMemWriter.java:165]: In option2, the records are not physically sorted. Thinking about the record read path, the [PMem+Disk] case is too complicate to support.

Comment (winningsix) [reply core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PMemWriter.java:165]: Yes, we can add TODO here.

File: core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java [hunk 7]
```diff
@@ -421,6 +421,72 @@ public void freePage(MemoryBlock page, MemoryConsumer consumer) {
     }
   }
 
+  /**
+   * @param size
+   * @return
+   */
+  public MemoryBlock allocatePMemPage(long size) {
```
[the remark this answers is not in the document]
Comment (yma11) [reply core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java:428]: Now both option2 and option3 doesn't use the logical memory acquire path any more. We just try to allocate the pages directly and if fail, we will fall back. The logical memory can't guarantee PMem allocation success finally.

File: core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java (hunk 2 above)
Comment (yma11) [reply core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java:452]: We allow the OOM happen. Handle the execption here means that allocatePMemPage returns a null value. I will update the code.

File: core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java (hunk 4 above)
Comment (yma11) [reply core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java:457]: I think we can. This is the logic use in Option3.

File: core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PMemWriter.java (hunk 5 above)
Comment (yma11) [reply core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PMemWriter.java:175]: yes

File: core/src/main/scala/org/apache/spark/memory/MemoryManager.scala [hunk 8]
```diff
@@ -69,6 +70,11 @@ private[spark] abstract class MemoryManager(
   protected[this] val extendedMemorySize = conf.get(MEMORY_EXTENDED_SIZE)
   extendedMemoryPool.incrementPoolSize((extendedMemorySize * 0.9).toLong)
 
+  private[memory] var _pMemPages = new ArrayBuffer[MemoryBlock];
+
+  private[memory] def pMemPages: ArrayBuffer[MemoryBlock] = {
```
[the remark this answers is not in the document]
Comment (yma11) [reply core/src/main/scala/org/apache/spark/memory/MemoryManager.scala:75]: But the page pool is an executor level pool. Where else you suggest?

File: core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (XinDongIntel) [reply core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java:428]: Option3 still use the original memory acquire path. There is no logic for Option3 to only try to acquire without actually allocation. Option2 seems don't need acquring withou allocate either. TaskMemoryManager is responsible for tracking the usage of every page, don't potentailly break the logic here.

File: core/src/main/scala/org/apache/spark/memory/MemoryManager.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (XinDongIntel) [reply core/src/main/scala/org/apache/spark/memory/MemoryManager.scala:75]: Maybe we can make the page reuse logic behind the PMemPageAllocator and init a executor level PMemPagePool to manage all the PMemPages.

[the remark this answers is not in the document]
Comment (yma11) [reply core/src/main/scala/org/apache/spark/memory/MemoryManager.scala:75]: We can think more details when implement the node-level PMem manager. That is the final feature we need.

**Comment by winningsix** [comment]
I didn't ask before. We're testing on an interleaved PMEM for DEV performance test? Just want to clarify this.

**Comment by yma11** [comment]
> I didn't ask before. We're testing on an interleaved PMEM for DEV performance test? Just want to clarify this.

The number I provided is using formal performance test environment(sr481 3 nodes cluster).

**Comment by winningsix** [comment]
> > I didn't ask before. We're testing on an interleaved PMEM for DEV performance test? Just want to clarify this.
> 
> 
> 
> The number I provided is using formal performance test environment(sr481 3 nodes cluster). The PMem is configured same as Plasma.

 Okay. So we configured PMEM by static partition.

File: core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PMemWriter.java [hunk 9]
```diff
@@ -104,12 +149,38 @@ private void updateLongArray(int numRecords, int position) {
...
+        }
+    }
+
+    public void clearAll() {
+        freeAllPMemPages();
```
[the remark this answers is not in the document]
Comment (yma11) [reply core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PMemWriter.java:171]: code updated.

File: core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (yma11) [reply core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java:428]: We can remove the PMem logic memory part in late PR.

File: core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PMemWriter.java (hunk 6 above)
Comment (yma11) [reply core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PMemWriter.java:166]: added.

File: core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java (hunk 1 above)
Comment (yma11) [reply core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java:437]: This is same as vanilla spark memory allocate.

**Comment by yma11** [comment]
Any more comments or we can merge it?

**Pull request merged**